lib/cylc/broadcast.py

```diff
@@ -15,89 +15,107 @@
+            keys, stuff, is_new = stuffs.pop()
+            if is_new:
+                stuffs.append((keys, stuff, False))
+                for key, value in stuff.items():
+                    if isinstance(value, dict):
+                        stuffs.append((keys + [key], value, True))
+            else:
+                for key, value in stuff.items():
+                    if not value:
+                        del stuff[key]
+                        prunes.append(keys + [key])
+        return prunes
+
+    def _addict(self, target, source):
+        """Recursively add source dict to target dict."""
         for key, val in source.items():
-            if isinstance( val, dict ):
+            if isinstance(val, dict):
                 if key not in target:
                     target[key] = {}
-                self.addict( target[key], val )
+                self._addict(target[key], val)
             else:
                 if source[key]:
                     target[key] = source[key]
                 elif key in target:
                     del target[key]
 
-    def put( self, namespaces, cycles, settings ):
-        """Add new broadcast settings, or prune newly unset ones"""
+    def put(self, point_strings, namespaces, settings):
+        """Add new broadcast settings.
+
+        Return a list of modified settings in the form:
+        [("20200202", "foo", {"command scripting": "true"}, ...]
+
+        """
+        modified_settings = []
         for setting in settings:
-            for cycle in cycles:
-                if cycle not in self.settings.keys():
-                    self.settings[cycle] = {}
+            for point_string in point_strings:
+                if point_string not in self.settings:
+                    self.settings[point_string] = {}
                 for namespace in namespaces:
-                    if namespace not in self.settings[cycle]:
-                        self.settings[cycle][namespace] = {}
-                    self.addict( self.settings[cycle][namespace], setting )
-        # prune any empty branches
-        self.prune( self.settings )
-
-        if self.get_dump() != self.last_settings:
-            current_time_string = get_current_time_string(
-                display_sub_seconds=True)
-            self.settings_queue.append(RecordBroadcastObject(
-                current_time_string, self.get_dump() ))
-            self.last_settings = self.settings
-
-        return ( True, 'OK' )
-
-    def get( self, task_id=None ):
-        # Retrieve all broadcast variables that target a given task ID.
+                    if namespace not in self.settings[point_string]:
+                        self.settings[point_string][namespace] = {}
+                    self._addict(
+                        self.settings[point_string][namespace], setting)
+                    modified_settings.append(
+                        (point_string, namespace, setting))
+
+        # Remove empty leaves
+        self._prune()
```

Comment:
I almost agree with you, but the trouble is, in parsec config files we make no distinction between "```key =``` (unset) and ```key = "   "``` (empty string), and it is arguably quite natural for users to expect that they can cancel a broadcast config item by setting it nothing, just as they can with the same item in the config file.  And pragmatically it doesn't matter because there are no config items that I'm aware of that can usefully take an empty string as a value (although conceivably that could change, I guess).